coverage/docs/design.md

```diff
@@ -0,0 +1,139 @@
+    - Stores in the XML format, that is used by TestGrid, and dump it in artifacts directory
+
+##Locally running presubmit and post-submit workflows
+Both workflows may be triggered locally in command line, as long as all the required flags are 
+supplied correctly. In addition, the following env var needs to be set:
+- JOB_TYPE (one of 'presubmit', 'postsubmit', 'periodic', 'local-presubmit')
+
+use 'local-presubmit' will run the presubmit workflow without posting result on github PR
+  
+##Covbot
+As mentioned in the presubmit workflow section, covbot is the short name for the robot github 
+account used to report code coverage change results. It can be created as a regular github 
+account - it does not need to be named covbot as that name is already taken on Github. It only need a 
+comment access to the repo it need to be run on. If the repo is private, it also need read access. 
+  
+After the robot account is created, download the github token and supply the path to the token 
+file to code coverage binary, as the value for parameter "github-token"
+
+#Usage with prow
+
+Prow can be used as the system to handle Github events mentioned in the two workflows. We can add a pre-submit prow job that is triggered by any new commit to a PR to run test coverage on the new build to compare it with the master branch and previous commit for pre-submit coverage. We can add a post-submit prow job that is triggered by merge events, to run test coverage when ever there is a merge on the target branch. 
+
+In addition, at the end of each workflow, prow copies the artifacts directory to gcs bucket
+
+
+##Prow Configuration File
+As mentioned earlier, we use configuration file to store repository specific information. Below is an example that contains the args that will be supplied to the coverage container
+```
+  - name: pull-knative-serving-go-coverage
+    labels:
+      preset-service-account: "true"
+    always_run: true
+    optional: true
+    decorate: true
+    clone_uri: "git@github.com:knative/serving.git"
+    ssh_key_secrets:
+    - ssh-knative
+    spec:
+      containers:
+      - image: gcr.io/knative-tests/test-infra/coverage:latest
+        imagePullPolicy: Always
+        command:
+        - "/coverage"
+        args:
+        - "--postsubmit-gcs-bucket=knative-prow"
+        - "--postsubmit-job-name=post-knative-serving-go-coverage"
+        - "--artifacts=$(ARTIFACTS)" # folder to store artifacts, such as the coverage profile
+        - "--profile-name=coverage_profile.txt"
+        - "--cov-target=./pkg/" # target directory of test coverage
+        - "--cov-threshold-percentage=50" # minimum level of acceptable presubmit coverage on a per-file level
+        - "--github-token=/etc/github-token/token"
+        volumeMounts:
+        - name: github-token
+          mountPath: /etc/github-token
+          readOnly: true
+      volumes:
+      - name: github-token
+        secret:
+          secretName: covbot-token
```

Comment:
We don't generally allow to expose credentials to jobs that run presubmit

Reply:
The covbot token is required to be unprivileged (like fejta-bot). I originally suggested a solution involving a code coverage job and a separate reporting component that has access to the covbot credentials, but I think this approach was chosen because it doesn't require writing code to watch for the coverage job to complete.

Reply:
it might be worth considering @krzyzacy's and @yutongz's reporting work in the future

Reply:
or spyglass